pkg/operator/controller/externaldns/rbac.go

```diff
@@ -0,0 +1,270 @@
+		}
+		return false, nil, err
+	}
+	return true, crb, nil
+}
+
+// createExternalDNSClusterRole creates the given cluster role
+func (r *reconciler) createExternalDNSClusterRole(ctx context.Context, desired *rbacv1.ClusterRole) error {
+	if err := r.client.Create(ctx, desired); err != nil {
+		return fmt.Errorf("failed to create externalDNS cluster role %s: %v", desired.Name, err)
+	}
+	r.log.Info("created externalDNS cluster role", "name", desired.Name)
+	return nil
+}
+
+// createExternalDNSClusterRoleBinding creates the given cluster role binding
+func (r *reconciler) createExternalDNSClusterRoleBinding(ctx context.Context, desired *rbacv1.ClusterRoleBinding) error {
+	if err := r.client.Create(ctx, desired); err != nil {
+		return fmt.Errorf("failed to create externalDNS cluster role binding %s: %v", desired.Name, err)
+	}
+	r.log.Info("created externalDNS cluster role binding", "name", desired.Name)
+	return nil
+}
+
+// updateExternalDNSClusterRole updates the cluster role with the desired state if the rules differ
+func (r *reconciler) updateExternalDNSClusterRole(ctx context.Context, current, desired *rbacv1.ClusterRole) (bool, error) {
+	changed, reason := externalDNSRoleRulesChanged(current.Rules, desired.Rules)
+	if !changed {
+		return false, nil
+	}
+
+	updated := current.DeepCopy()
+	// no complicated resets of specific rules are used
+	// if there is change: reset all
+	updated.Rules = desired.Rules
+	if err := r.client.Update(ctx, updated); err != nil {
+		return false, err
+	}
+	r.log.Info("updated externalDNS cluster role", "name", updated.Name, "reason", reason)
+	return true, nil
+}
+
+// updateExternalDNSClusterRoleBinding updates the cluster role binding with the desired state if the role or subject changed
+func (r *reconciler) updateExternalDNSClusterRoleBinding(ctx context.Context, current, desired *rbacv1.ClusterRoleBinding) (bool, error) {
+	updated := current.DeepCopy()
+
+	changed, reason := externalDNSRoleBindingChanged(current, desired, updated)
+	if !changed {
+		return false, nil
+	}
+
+	if err := r.client.Update(ctx, updated); err != nil {
+		return false, err
+	}
+	r.log.Info("updated externalDNS cluster role binding", "name", updated.Name, "reason", reason)
+	return true, nil
+}
+
+// externalDNSRoleRulesChanged returns true if the contents of the rules changed
+// order of apigroups, resources, verbs does not matter
+func externalDNSRoleRulesChanged(current, expected []rbacv1.PolicyRule) (bool, string) {
+	currentRuleMap := buildPolicyRuleMap(current)
+	expectedRuleMap := buildPolicyRuleMap(expected)
+
+	for resource, currVerbs := range currentRuleMap {
+		if expVerbs, found := expectedRuleMap[resource]; found {
+			if !equalStringSliceContent(currVerbs, expVerbs) {
+				return true, fmt.Sprintf("policy rule resource %q has %v vebs but expected to have %v", resource, currVerbs, expVerbs)
+			}
+		} else {
+			return true, fmt.Sprintf("not expected policy rule resource %q", resource)
+		}
```

Comment:
What if expected is a proper superset of current?  For example, should the following test case pass?  
```go
		{
			name: "current is missing a rule",
			inputCurrent: []rbacv1.PolicyRule{},
			inputExpected: []rbacv1.PolicyRule{
				{
					APIGroups: []string{"networking.k8s.io"},
					Resources: []string{"ingresses"},
					Verbs:     []string{"get", "list", "watch"},
				},
			},
			expectChanged: true,
		},
```

Reply:
There was a bug in the comparison. I found it in the later PRs, Andy did find it during his review and you did too. We are all good :)
I changed the implementation of this function to get 2 maps with sorted values (slices) and I compare them using `go-cmp`.

Reply:
If we need to ignore ordering, it might be worth using go-cmp's `SortSlices` transformer so that you don't need `buildSortedPolicyRuleMap`.